src/Bicep.Core/Resources/ResourceTypeReferenceComparer.cs

```diff
@@ -0,0 +1,25 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Bicep.Core.Resources
+{
+    public class ResourceTypeReferenceComparer : IEqualityComparer<ResourceTypeReference>
+    {
+        public static IEqualityComparer<ResourceTypeReference> Instance { get; }
+            = new ResourceTypeReferenceComparer();
+
+        public bool Equals(ResourceTypeReference x, ResourceTypeReference y)
+            => StringComparer.OrdinalIgnoreCase.Equals(x.Namespace, y.Namespace) &&
+            x.Types.Length == y.Types.Length &&
+            Enumerable.SequenceEqual(x.Types, y.Types, StringComparer.OrdinalIgnoreCase) &&
+            StringComparer.OrdinalIgnoreCase.Equals(x.ApiVersion, y.ApiVersion);
+
+        public int GetHashCode(ResourceTypeReference obj)
```

Comment:
>GetHashCode [](start = 19, length = 11)

R# has an interesting approach for generating GetHashCode function that intentionally forces integer overflow in an unchecked scope. If we're going to be putting these in dictionaries a lot, we should take a look at [https://stackoverflow.com/questions/102742/why-is-397-used-for-resharper-gethashcode-override](https://stackoverflow.com/questions/102742/why-is-397-used-for-resharper-gethashcode-override).

Reply:
Yeah, may as well optimize this as it'll be used reasonably frequently. I might let you add this in a later PR, as I don't feel like installing R#.

Reply:
No need to install it - can just modify the code from SO

---
In reply to: [493933232](https://github.com/Azure/bicep/pull/553#discussion_r493933232) [](ancestors = 493933232)